lib/mlir/Dialect/Clift/IR/ImportModel.cpp

```diff
@@ -0,0 +1,599 @@
+    const clift::ValueType T = getQualifiedType(ModelType,
+                                                /*RequireComplete=*/true);
+    if (T and not processIncompleteTypes()) {
+      IncompleteTypes.clear();
+      return nullptr;
+    }
+    return T;
+  }
+
+private:
+  mlir::BoolAttr getBool(bool const Value) {
+    return mlir::BoolAttr::get(Context, Value);
+  }
+
+  mlir::BoolAttr getFalse() { return getBool(false); }
+
+  template<typename T, typename... ArgTypes>
+  T make(const ArgTypes &...Args) {
+    if (EmitError and failed(T::verify(EmitError, Args...)))
+      return {};
+    return T::get(Context, Args...);
+  }
+
+  static clift::PrimitiveKind
+  getPrimitiveKind(const model::PrimitiveTypeKind::Values K) {
+    switch (K) {
+    case model::PrimitiveTypeKind::Void:
+      return clift::PrimitiveKind::VoidKind;
+    case model::PrimitiveTypeKind::Generic:
+      return clift::PrimitiveKind::GenericKind;
+    case model::PrimitiveTypeKind::PointerOrNumber:
+      return clift::PrimitiveKind::PointerOrNumberKind;
+    case model::PrimitiveTypeKind::Number:
+      return clift::PrimitiveKind::NumberKind;
+    case model::PrimitiveTypeKind::Unsigned:
+      return clift::PrimitiveKind::UnsignedKind;
+    case model::PrimitiveTypeKind::Signed:
+      return clift::PrimitiveKind::SignedKind;
+    case model::PrimitiveTypeKind::Float:
+      return clift::PrimitiveKind::FloatKind;
+    default:
+      revng_abort();
+    }
+  }
+
+  clift::ValueType getPrimitiveType(const model::PrimitiveType &ModelType,
+                                    const bool Const) {
+    return make<clift::PrimitiveType>(getPrimitiveKind(ModelType
+                                                         .PrimitiveKind()),
+                                      ModelType.Size(),
+                                      getBool(Const));
+  }
+
+  RecursiveCoroutine<clift::TypeDefinition>
+  getTypeAttribute(const model::CABIFunctionType &ModelType) {
+    DefinitionGuard Guard(*this, ModelType.ID());
+    if (not Guard) {
+      if (EmitError)
+        EmitError() << "Recursive definition of CABIFunctionType "
+                    << ModelType.ID();
+      rc_return nullptr;
+    }
+
+    AttributeVector<clift::FunctionArgumentAttr> Args;
+    Args.reserve(ModelType.Arguments().size());
+
+    for (const model::Argument &Argument : ModelType.Arguments()) {
+      const auto Type = rc_recur getQualifiedType(Argument.Type());
+      if (not Type)
+        rc_return nullptr;
+      const llvm::StringRef Name = Argument.OriginalName();
+      const auto Attribute = make<clift::FunctionArgumentAttr>(Type, Name);
+      if (not Attribute)
+        rc_return nullptr;
+      Args.push_back(Attribute);
+    }
+
+    const auto ReturnType = rc_recur getQualifiedType(ModelType.ReturnType());
+    if (not ReturnType)
+      rc_return nullptr;
+
+    rc_return make<clift::FunctionAttr>(ModelType.ID(),
+                                        ModelType.OriginalName(),
```

Comment:
I'm putting this comment here, but it applies to any other place dealing with `OriginalName` vs `CustomName`.

Basically, this is linked to the issue of uniqueness of names.
`CustomName`s are already guaranteed to be uniqued by the Model where relevant.
So one thing we could do is: switch back to requiring uniqueness, and use `CustomName()` everywhere instead of `OriginalName()`, to get uniqueness from the model for free. If anybody ever writes another frontend, they'll have to deal with uniqueness in some other way, but for us for now it's not a problem.
It seems to me like a path of low impedance, not necessarily the best design in the long term.

In any case, whatever we decide about uniqueness of names, the clift type importer should be self-consistent and either always use `OriginalName()` or `CustomName()` everywhere. Now some places use `OriginalName`, some others `CustomName`. Any specific reason for that?

Reply:
> Any specific reason for that?

No, I don't think so. Let's pick one and I'll change them to be consistent.

Reply:
After more discussion on Mattermost, these were all changed to use `name()` and `entryName()` for enum fields.